add log opts flag to pass in logging options

wlan0 opened the pull request:
As per the discussions on this -- https://github.com/docker/docker/pull/11485/files

I'm dividing #11485 into two Pull requests, one for log-opts and one for `max-size` and `max-file` options.

This is the PR for log-opts.

@ibuildthecloud
@tiborvass 
@LK4D4

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "logopts" git@github.com:wlan0/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

ibuildthecloud commented on runconfig/parse.go:
```diff
@@ -375,6 +382,23 @@ func convertKVStringsToMap(values []string) map[string]string {
+	if loggingDriver == "none" && len(loggingOpts) > 0 {
+		return map[string]string{}, fmt.Errorf("Invalid logging opts for driver %s", loggingDriver)
+	}
+	//validation step
```
Can you remove the commented lines?

wlan0 replied:
@ibuildthecloud I wanted to enforce a validation step at this point for the log options. The commented lines have a pattern for validating log options.

thaJeztah replied:
Perhaps just a `TODO`? The commented code looks trivial

LK4D4 replied:
Yeah, `TODO` should be ok.

LK4D4 commented on opts/opts.go:
```diff
@@ -24,6 +24,10 @@ func ListVar(values *[]string, names []string, usage string) {
 	flag.Var(newListOptsRef(values, nil), names, usage)
 }
 
+func MapVar(values *map[string]string, names []string, usage string) {
```
I really think that pointer isn't necessary here.

wlan0 replied:
@LK4D4 You're right, theres no need for one. I was trying to be consistent with the rest of the types. Changed it just the map type now

LK4D4 commented:
Shouldn't we pass this to logging drivers?

LK4D4 commented on opts/opts.go:
```diff
@@ -54,6 +58,31 @@ type ListOpts struct {
+}
+
+func (opts *MapOpts) Set(value string) error {
+	vals := strings.Split(value, "=")
```
Maybe SplitN with 2?

wlan0 replied:
That would allow one to specify an option like `--map-opt key=value1=value2` I cant think of a use case for something like that. Are there any potential benefits of doing that?

wlan0 replied:
Just checked runconfig/parse.go, and it does splitN with n=2. For consistency between `docker run` and `docker -d`, i'll make it splitN

LK4D4 commented:
Also tests for MapOpts needed.

thaJeztah commented:
Thanks for splitting it, @wlan0!

LK4D4 commented on opts/opts_test.go:
```diff
@@ -64,6 +65,15 @@ func TestListOpts(t *testing.T) {
+func TestMapOpts(t *testing.T) {
+	tmpMap := make(map[string]string)
+	o := newMapOpt(tmpMap)
+	o.Set("max-size=1")
```
makes sense to test 2 keys

LK4D4 commented:
Why do you plan validation for daemon and not for run?

wlan0 commented:
@LK4D4  Addressed all your comments I think.

Good catch btw, we need validation for docker daemon as well. 

Now there are,
- Tests for MapOpts
- Validation for docker daemon
- other minor code change.

LK4D4 commented on opts/opts.go:
```diff
@@ -130,6 +172,15 @@ func (opts *ListOpts) Len() int {
 
+func ValidateLogOpts(val string) (string, error) {
+	allowedKeys := map[string]string{}
+	vals := strings.Split(val, "=")
```
Is this basically `strings.Contains` check?

wlan0 replied:
yes

LK4D4 replied:
Then I think better to use `strings.Contains` :)

wlan0 replied:
@LK4D4 This cannot be `strings.Contains`. `strings.Contains` will validate `some-string=allowed-string`, whereas we only want `allowed-string=some-string` to be valid.

LK4D4 replied:
ah, okay.

LK4D4 commented on opts/opts_test.go:
```diff
@@ -64,6 +65,31 @@ func TestListOpts(t *testing.T) {
+	if tmpMap["max-size"] != "1" {
+		t.Errorf("max-size = %s != 1", tmpMap["max-size"])
+	}
+	if o.Set("dummy-val=3") == nil {
```
Why it is invalud?

tiborvass commented:
@wlan0 what is this PR supposed to do? Just introduce log-opt and passing it to the logging driver and that's it? If that's the case I would remove references to `max-size` etc. in the docs.

wlan0 commented:
@tiborvass Nice catch, thanks. I removed references to `max-size` and `max-file`

tiborvass commented:
@wlan0 please add some tests to make sure a dummy logging driver gets some dummy configs.

tiborvass commented on daemon/config.go:
```diff
@@ -75,6 +75,8 @@ func (config *Config) InstallFlags() {
 	config.Ulimits = make(map[string]*ulimit.Ulimit)
 	opts.UlimitMapVar(config.Ulimits, []string{"-default-ulimit"}, "Set default ulimits for containers")
 	flag.StringVar(&config.LogConfig.Type, []string{"-log-driver"}, "json-file", "Default driver for container logs")
+	config.LogConfig.Config = make(map[string]string)
```
this could be put in the newMapOpt constructor. Just check if map is nil, and create it if it's nil in the constructor.

wlan0 replied:
In that case, I'd have to pass a pointer. Which was my original implementation of this. But, I changed it to this approach after comments from @LK4D4

LK4D4 replied:
Why you need pass pointer?

wlan0 replied:
@LK4D4 map[string]string is a reference type. It is a pointer to the underlying map. When you pass it by value, the pointer value gets copied. If i allocate it after passing it by value, the copy gets allocated, not the original map. 
https://play.golang.org/p/HB4qUGYoGE

LK4D4 replied:
I think maybe better to instantiate such things in `docker/daemon.go:init()` where empty daemon.Config is created

wlan0 replied:
ok

wlan0 commented:
@tiborvass There are already tests with some dummy log configs being passed in. Do you mean integration/-cli tests?

wlan0 commented:
@LK4D4 I've updated the code to initialize it in `docker/daemon.go:init()`

LK4D4 commented:
Code okay for me. I'm eager to see it's usage.
@tiborvass WDYT about adding it to syslog right here?

ibuildthecloud commented:
FYI, Syslog implementation using `--log-opts` in #12391

tiborvass commented:
@ibuildthecloud the PR you referenced doesn't use `--log-opt` or am I missing something?

@LK4D4 I think I'd prefer in a separate PR, so we can merge this, but I don't really mind.

tiborvass commented on daemon/config.go:
```diff
@@ -75,6 +75,7 @@ func (config *Config) InstallFlags() {
 	config.Ulimits = make(map[string]*ulimit.Ulimit)
 	opts.UlimitMapVar(config.Ulimits, []string{"-default-ulimit"}, "Set default ulimits for containers")
 	flag.StringVar(&config.LogConfig.Type, []string{"-log-driver"}, "json-file", "Default driver for container logs")
+	opts.LogOptsVar(config.LogConfig.Config, []string{"-log-opt"}, "Set log driver options")
```
@LK4D4 why not use `opts.NewListOpts` with opts.ValidateEnv like --label ?
Also there are already functions like convertKVStringsToMap that we can reuse.

LK4D4 replied:
Have no idea :/ Wanna carry this?

ibuildthecloud commented:
Lol, wrong PR, too many late night drunken pull requests.  Here's the right one https://github.com/docker/docker/pull/12668

wlan0 committed:
Add log opts flag to pass in logging options

Signed-off-by: wlan0 <sidharthamn@gmail.com>

icecrime commented on opts/opts.go:
```diff
@@ -54,6 +62,40 @@ type ListOpts struct {
+}
+
+func (opts *MapOpts) String() string {
+	return fmt.Sprintf("%v", map[string]string((opts.values)))
```
Nit: extra pair of parenthesis.

icecrime commented on opts/opts.go:
```diff
@@ -130,6 +172,15 @@ func (opts *ListOpts) Len() int {
+func ValidateLogOpts(val string) (string, error) {
+	allowedKeys := map[string]string{}
+	vals := strings.Split(val, "=")
+	if allowedKeys[vals[0]] != "" {
```
Sorry I don't get this line: `allowedKeys` is always a newly constructed map, so what's the point of this test?

wlan0 replied:
`allowedKeys` is intended to contain the keys. Right now, there are no keys defined. Once someone adds a new key,  they will have to add that key into this map, otherwise, the validator will also reject that key.

icecrime replied:
Ooooh ok I see, soryr.

icecrime commented:
@wlan0 Can you please rebase?

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "logopts" git@github.com:wlan0/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

icecrime commented:
Thanks, code LGTM.

wlan0 commented:
@icecrime rebased.

jfrazelle commented:
ah man so sorry needs rebase then you can ping me for review and merge thanks!

wlan0 commented:
@jfrazelle Rebased :)

jfrazelle commented:
LGTM thanks!

wlan0 commented:
Thanks all!

ahmetalpbalkan commented:
@LK4D4 @wlan0 so I can see that `--log-opt` is added to the daemon command. Does this PR also make  `--log-opts`  available on `docker run` because that's another place we configure logging drivers. (I can't see it documented in this PR.)

If one of you can provide more context I can start working on plumbing LoggingOpts to the logdrivers factory using logging.Context.

ahmetalpbalkan commented on daemon/config.go:
```diff
@@ -80,6 +80,7 @@ func (config *Config) InstallFlags() {
 	opts.UlimitMapVar(config.Ulimits, []string{"-default-ulimit"}, "Set default ulimits for containers")
 	flag.StringVar(&config.LogConfig.Type, []string{"-log-driver"}, "json-file", "Default driver for container logs")
 	flag.BoolVar(&config.Bridge.EnableUserlandProxy, []string{"-userland-proxy"}, true, "Use userland proxy for loopback traffic")
+	opts.LogOptsVar(config.LogConfig.Config, []string{"-log-opt"}, "Set log driver options")
```
@wlan0 isn't this supposed to be `&config.LogConfig.Config`  (not sure, just reading the code)

wlan0 replied:
@ahmetalpbalkan Config is a map, its already a reference type
